### PR TITLE
Caching forts, saving + loading cache, spin cached forts

### PR DESCRIPTION
**Pull request by cryzies**
Last thread: https://github.com/j-e-k/poketrainer/pull/397

Same thing, lot's of bug fixes I discovered when fixing conflicts and building the cache. Should work, should....

Things to note in config:

```
"ENABLE_CACHING" :true,
"USE_CACHED_FORTS" : false,
"CACHED_FORTS_SORTED" : false
```

**ENABLE_CACHING** is the master switch, no caching methods will run if this is set to false
**USE_CACHED_FORTS** should be set to false on your first run, it will run as normal and cache forts you'd normally visit. Set this to true after your "Cached Forts: x" output is stable, usually 15-60minutes depending on your step and proximity.
**CACHED_FORTS_SORTED** was added for large pokestop counts, if set to true, the cache is already sorted and the code won't run a greedy algorithm to sort the pokestops. If false, the cache will be sorted and saved to file. This will probably be removed in the future, I sorted 400 stops in less than a few seconds, so it's not worth the confusion imo.

### CODE REVIEW

**Comment by cryzies** [comment]
201latlon

Example how the routing is done with 201 forts.

File: pgoapi/pgoapi.py [hunk 1]
```diff
@@ -178,6 +183,14 @@ def __init__(self, config):
         self.should_catch_pokemon = config.get("CAPTURE", {}).get("CATCH_POKEMON", True)
         self.max_catch_attempts = config.get("CAPTURE", {}).get("MAX_CATCH_ATTEMPTS", 10)
 
+        self.new_forts = []
+        self.cache_filename = './cache/cache ' + str(config["location"]) + str(self.STAY_WITHIN_PROXIMITY)
```
Comment (destiny117) [inline pgoapi/pgoapi.py]: i think `config["location"]` could lead to some problems here:
- First you should probably use `config.get("location", "unknown")` or something
- Next there can be issues file prohibited characters for filenames on different OSs, maybe use a hash on the location or something like that?

**Comment by destiny117** [comment]
I haven't checked everything, but as far as i can tell this should not change the bots default behaviour so should be safe to merge :+1: 

noted one small thing i saw that might cause trouble and you should check

**Comment by j-e-k** [comment]
Let's get this merged. @destiny117 feel free to do the merge once you see your concerns fixed.

Thanks for the contribution 😄

**Comment by stolencatkarma** [comment]
care to update the readme as well with the new config options? Maybe you already grabbed it.

File: pgoapi/pgoapi.py (hunk 1 above)
Comment (cryzies) [inline pgoapi/pgoapi.py]: @destiny117 hmmm, copied the syntax from the pgoapi and pokecli https://github.com/j-e-k/poketrainer/blob/master/pokecli.py#L123

Should error there before it gets to my code :P

**Comment by destiny117** [comment]
yes, a small readmy update and a gitignore for the cache directory :+1:

**Comment by cryzies** [comment]
Have made the changes needed, won't commit until I can test it though :/

**Comment by j-e-k** [comment]
Hey @cryzies Can you test this now that the bot works :)

**Comment by cryzies** [comment]
@j-e-k should I worry about the conficts?

**Comment by j-e-k** [comment]
Probably need to rebase to get the unknown 6 fixes. Feel free to catch me on slack (I am jekirl) if you need help

**Comment by infinitewarp** [comment]
@cryzies : FWIW, I just rebased your changes, fixed the conflicts, and created a new PR here: https://github.com/j-e-k/poketrainer/pull/424

Thoughts?

**Comment by cryzies** [comment]
@infinitewarp thanks for the work! Was working on it as well but didn't see, would've saved me some time :d, anyways rebased and added gitignore for /cache and updated readme with parameters

**Comment by j-e-k** [comment]
Anything else holding up this merge or are we good to go?

**Comment by infinitewarp** [comment]
I've been playing a bit with this branch, and I have at least one consistent problem: Whenever I have `"CACHED_FORTS_SORTED" : false` in my config, the following exception is raised:

```
2016-08-07 18:39:35,804 [    pgoapi] [ INFO] ==== CACHING MODE: ROUTE+SPIN CACHED FORTS ====
2016-08-07 18:39:36,412 [    pgoapi] [ INFO] Player Info: [redacted], Currencies: POKECOIN: 50,STARDUST: 237569, Pokemon Caught in this run: 0
2016-08-07 18:39:36,413 [    pgoapi] [ INFO] Cache is unsorted, sorting now...
2016-08-07 18:39:36,413 [   pokecli] [ERROR] Error in main loop list index out of range, restarting at location: [redacted]
Traceback (most recent call last):
  File "pokecli.py", line 166, in main
    api.main_loop()
  File "/Users/infinitewarp/Documents/code/poketrainer/pgoapi/pgoapi.py", line 1299, in main_loop
    self.spin_all_cached_forts()
  File "/Users/infinitewarp/Documents/code/poketrainer/pgoapi/pgoapi.py", line 1223, in spin_all_cached_forts
    destinations = self.sort_cached_forts()
  File "/Users/infinitewarp/Documents/code/poketrainer/pgoapi/pgoapi.py", line 1183, in sort_cached_forts
    tempsorted = [copy.deepcopy(tempallcached[0])] # the final list to copy to cache
IndexError: list index out of range
```

**Comment by cryzies** [comment]
@infinitewarp, I see where the problem is coming from, it's trying to sort an empty cache. Did you first run it with only enable_caching as true and use_cached_forts as false? It needs to run normally and create the cache file, then after it creates the cache file you can do use_cached_forts as true. I'll put in some lines to prevent this from happening.

**Comment by sontek** [comment]
I'd merge this if you fixed this:

```
should be set to false on your first run
```

I think we should provide this for them by having a threshold of required forts before we start caching (or make it configurable) and just using a heuristic on how stable the fort path is and if its ready to be cached?

**Comment by sontek** [comment]
@cryzies I think if you fixed that it would also fix the issue @infinitewarp ran into

**Comment by infinitewarp** [comment]
@cryzies Good catch. As I've been switching between branches and trying out different features, I have also been purging untracked files to keep my environment clean, and that included the cache directory in this case.

If you can put in a fix to check for empty cache on first run, that would be great! I'm sure I wouldn't be the only person to hit that issue.

**Comment by cryzies** [comment]
Pushed the fix for using cached forts when there are none, it will automatically switch to caching mode and run normally.

**Comment by infinitewarp** [comment]
I haven't pulled your latest commit yet to fix that one issue, but otherwise the new code on this branch has been very stable for me. I gathered the visited fort locations for my latest run, and although the picture isn't as pretty as your example and there are some weird offshoots, it certainly looks like it's trying a better pattern than before.

map

**Comment by infinitewarp** [comment]
Latest pull looks good to me. Finally time to merge. 🎉

File: README.md [hunk 2]
```diff
@@ -57,6 +57,9 @@ Below the accounts you can change options in the `default` section. If you need
    * `SKIP_VISITED_FORT_DURATION` [Experimental] Avoid a fort for a given number of seconds
      * Setting this to 500 means avoid a fort for 500 seconds before returning, (Should be higher than 300 to have any effect). This will let the bot explore a bigger area.
    * `SPIN_ALL_FORTS` [Experimental] will try to route using google maps(must have key) to all visible forts, if `SKIP_VISITED_FORT_DURATION` is set high enough, you may roam around forever.
+   * `ENABLE_CACHING` is the master switch, no caching methods will run if this is set to false
+   * `USE_CACHED_FORTS` should be set to false on your first run, it will run as normal and cache forts you'd normally visit. Set this to true after you've cached enough forts or "Cached Forts: x" output is stable, around 10 for a 1500 proximity, 100 step configuration, wait longer if needed.
```
Comment (sontek) [inline README.md]: need to update the docs here since this isn't required to be ran with it off at first anymore

**Comment by j-e-k** [comment]
Yay! 🎉

**Pull request merged**